# ensure-forward-ref-using-ref

## Rule category

Correctness.

## What it does

Requires that components wrapped with `forwardRef` must have a `ref` parameter.

This rule checks all React components using `forwardRef` and verifies that there is a second parameter.

## Why is this bad?

Omitting the `ref` argument is usually a bug, and components not using `ref` don't need to be wrapped by `forwardRef`.

## Examples

### Failing

```tsx twoslash
import React from "react";

// @warn: Missing 'ref' parameter
const Example = React.forwardRef((props) => {
  return <button />;
});
```

### Passing

```tsx twoslash
import React from "react";

// @annotate: The 'ref' parameter is being used
const Example = React.forwardRef<HTMLButtonElement>((props, ref) => {
  return <button ref={ref} />;
});
```

## Further Reading

- [react.dev: forwardRef](https://react.dev/reference/react/forwardRef)
