# no-useless-fragment

## Rule category

Correctness.

## What it does

Prevents the use of useless `fragment` components or `<>` syntax.

## Why is this bad?

A fragment is redundant if it contains only one child, or if it is the child of a html element, and is not a [keyed fragment](https://react.dev/reference/react/Fragment#caveats).

And, it adds unnecessary visual noise to the code.

## Examples

### Failing

```tsx twoslash
import React, { Fragment } from "react";

function Example() {
// @warn: Useless fragment
  return <Fragment>foo</Fragment>;
}
```

### Passing

```tsx twoslash
import React, { Fragment } from "react";

function Example() {
  return (
    <Fragment>
      <div>foo</div>
      <div>bar</div>
      <div>baz</div>
    </Fragment>
  );
}
```
