# no-useless-fragment

## Rule category

Correctness.

## What it does

Prevents the use of useless `fragment` components or `<>` syntax.

## Why is this bad?

A fragment is redundant if it contains only one child, or if it is the child of a html element, and is not a [keyed fragment](https://react.dev/reference/react/Fragment#caveats).

And, it adds unnecessary visual noise to the code.

## Examples

### Failing

```tsx twoslash
import React, { Fragment } from "react";

function Example() {
  return <Fragment>foo</Fragment>;
  //     ^^^^^^^^^^^^^^^^^^^^^^^^
  //     - Fragments containing a single element are usually unnecessary.
}
```

### Passing

```tsx twoslash
import React, { Fragment } from "react";

function Example() {
  return (
    <Fragment>
      <div>foo</div>
      <div>bar</div>
      <div>baz</div>
    </Fragment>
  );
}
```

## Note

[This rule always allows single expressions in a fragment](https://github.com/Rel1cx/eslint-react/pull/188), even though this was previously controlled by the `allowExpressions` option. This is useful in
places like Typescript where `string` does not satisfy the expected return type
of `JSX.Element`. A common workaround is to wrap the variable holding a string
in a fragment and expression.

### Examples of correct code for single expressions in fragments:

```tsx twoslash
import React from "react";

function Example() {
  const foo = "bar";
  return <>{foo}</>;
}

function AnotherExample() {
  const foo = "bar";
  return (
    <Fragment>
      {foo}
    </Fragment>
  );
}
```
